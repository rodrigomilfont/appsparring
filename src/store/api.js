export { fetchSearch } from './search/endpoint';

export function checkStatus(response) {
  console.log('response: ', response);

  console.log('status: ', response.status);

  return response;
}
