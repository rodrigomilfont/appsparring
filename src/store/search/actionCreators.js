import {
  SET_SEARCH_TERM,
  REQUEST_SEARCH,
  RECEIVE_SEARCH,
  FAILED_SEARCH,
} from './actions';

export function setSearchTerm(searchTerm) {
  return {
    type: SET_SEARCH_TERM,
    payload: searchTerm,
  };
}

export function requestSearch(searchTerm) {
  return {
    type: REQUEST_SEARCH,
    searchTerm,
  };
}

export function receiveSearch(searchTerm, json) {
  return {
    type: RECEIVE_SEARCH,
    searchTerm,
    results: json,
    receiveAt: Date.now(),
  };
}

export function failedSearch(searchTerm, error) {
  return {
    type: FAILED_SEARCH,
    searchTerm,
    error,
  };
}

export function fetchSearch(searchTerm) {
  return (dispatch, getState, api) => {
    dispatch(requestSearch(searchTerm));
    return api
      .fetchSearch(searchTerm)
      .then(api.checkStatus)
      .then(response => {
        dispatch(receiveSearch(searchTerm, response.data));
      })
      .catch(error => {
        dispatch(failedSearch(searchTerm, error));
      });
  };
}
