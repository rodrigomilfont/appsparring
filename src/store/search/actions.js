export const SET_SEARCH_TERM = 'SET_SEARCH_TERM';
export const ADD_API_DATA = 'ADD_API_DATA';

export const REQUEST_SEARCH = 'REQUEST_SEARCH';
export const RECEIVE_SEARCH = 'RECEIVE_SEARCH';
export const FAILED_SEARCH = 'FAILED_SEARCH';
