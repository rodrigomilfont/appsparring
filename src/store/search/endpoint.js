import axios from 'axios';

export const fetchSearch = searchTerm =>
  axios({
    method: 'get',
    url: `https://api.mercadolibre.com/sites/MLA/search?q=${searchTerm}`,
    headers: { 'X-Coma-With': 'XMLHttpRequest' },
  });
