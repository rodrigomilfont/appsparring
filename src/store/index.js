import { createStore, compose, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from './reducers';
import * as api from './api';

const store = createStore(
  rootReducer,
  compose(
    applyMiddleware(thunk.withExtraArgument(api)),
    typeof window === 'object' &&
    typeof window.devToolsExtension !== 'undefined'
      ? window.devToolsExtension()
      : f => f,
  ),
);

export default store;
