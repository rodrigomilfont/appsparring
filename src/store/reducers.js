import { combineReducers } from 'redux';
import search from './search/reducers';

export default combineReducers({
  search,
});
