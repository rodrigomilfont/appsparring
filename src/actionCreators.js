import axios from 'axios';
import { SET_SEARCH_TERM, ADD_API_DATA } from './actions';

export function setSearchTerm(searchTerm) {
  return {
    type: SET_SEARCH_TERM,
    payload: searchTerm,
  };
}

export function addAPIData(apiData) {
  return {
    type: ADD_API_DATA,
    payload: apiData,
  };
}

export function getAPIDetails(searchTerm) {
  return dispatch => {
    axios
      .get(`https://api.mercadolibre.com/sites/MLA/search?q=${searchTerm}`)
      // .get(`https://swapi.co/api/people/?search=${searchTerm}`)
      .then(response => {
        dispatch(addAPIData(response.data));
      })
      .catch(error => {
        console.error('axios error', error); // eslint-disable-line no-console
      });
  };
}
