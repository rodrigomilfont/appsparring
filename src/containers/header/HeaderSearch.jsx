import React from 'react';
import { Link, withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import './index.css';
import { setSearchTerm, fetchPosts } from '../../actionCreators';

class HeaderSearch extends React.Component {
  constructor(props) {
    super(props);

    this.handleSubmit = event => {
      event.preventDefault();
      this.props.history.push(`/items?search=${this.props.searchTerm}`);
      this.props.fetchPostAsync(this.props.searchTerm);
    };
  }

  render() {
    return (
      <div className="box-search">
        <form onSubmit={this.handleSubmit}>
          <input
            onChange={this.props.handleSearchTermChange}
            value={this.props.searchTerm}
            type="text"
            placeholder="Search"
            tabIndex="0"
          />
        </form>
        <Link
          to={`/items?search=${this.props.searchTerm}`}
          className="bt-search">
          Buscar
        </Link>
      </div>
    );
  }
}

HeaderSearch.propTypes = {
  fetchPostAsync: PropTypes.func,
  handleSearchTermChange: PropTypes.func,
  searchTerm: PropTypes.string,
  history: PropTypes.shape({
    push: PropTypes.func,
  }),
};

HeaderSearch.defaultProps = {
  fetchPostAsync: function noop() {},
  handleSearchTermChange: function noo() {},
  searchTerm: '',
  history: {
    push: function noo() {},
  },
};

const mapStateToProps = state => {
  const { searchTerm } = state;
  return {
    searchTerm,
  };
};

const mapDispatchToProps = dispatch => ({
  handleSearchTermChange(event) {
    dispatch(setSearchTerm(event.target.value));
  },
  fetchPostAsync(searchTerm) {
    dispatch(fetchPosts(searchTerm));
  },
});

export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(HeaderSearch),
);
