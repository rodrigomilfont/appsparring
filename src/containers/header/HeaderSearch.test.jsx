import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import store from '../../store';
import { setSearchTerm, failedSearch } from '../../store/search/actionCreators';
import HeaderSearch, {
  Unwrapped as UnwrappedHeaderSearch,
} from '../../containers/header/HeaderSearch';

Enzyme.configure({ adapter: new Adapter() });

function setup() {
  const wrapper = mount(
    <Provider store={store}>
      <MemoryRouter>
        <HeaderSearch />
      </MemoryRouter>
    </Provider>,
  );
  return {
    wrapper,
  };
}

describe('Unwrapped', () => {
  it('renders without crashing', () => {
    const wrapper = shallow(<UnwrappedHeaderSearch />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Connected HeaderSearch', () => {
  const searchTerm = 'Goku';

  it('change search term', () => {
    store.dispatch(setSearchTerm(searchTerm));

    const { wrapper } = setup();
    expect(wrapper.find('input').props().value).toEqual(searchTerm);
  });
});
