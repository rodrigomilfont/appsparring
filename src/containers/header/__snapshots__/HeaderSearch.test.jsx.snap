// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unwrapped renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <HeaderSearch
    fetchSearchAsync={[Function]}
    handleSearchTermChange={[Function]}
    history={
      Object {
        "push": [Function],
      }
    }
    searchTerm=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <form
          onSubmit={[Function]}
        >
          <input
            onChange={[Function]}
            placeholder="Search"
            tabIndex="0"
            type="text"
            value=""
          />
        </form>,
        <Link
          className="bt-search"
          replace={false}
          to="/items?search="
        >
          Buscar
        </Link>,
      ],
      "className": "box-search",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <input
            onChange={[Function]}
            placeholder="Search"
            tabIndex="0"
            type="text"
            value=""
          />,
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onChange": [Function],
            "placeholder": "Search",
            "tabIndex": "0",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Buscar",
          "className": "bt-search",
          "replace": false,
          "to": "/items?search=",
        },
        "ref": null,
        "rendered": "Buscar",
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <form
            onSubmit={[Function]}
          >
            <input
              onChange={[Function]}
              placeholder="Search"
              tabIndex="0"
              type="text"
              value=""
            />
          </form>,
          <Link
            className="bt-search"
            replace={false}
            to="/items?search="
          >
            Buscar
          </Link>,
        ],
        "className": "box-search",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <input
              onChange={[Function]}
              placeholder="Search"
              tabIndex="0"
              type="text"
              value=""
            />,
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "onChange": [Function],
              "placeholder": "Search",
              "tabIndex": "0",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Buscar",
            "className": "bt-search",
            "replace": false,
            "to": "/items?search=",
          },
          "ref": null,
          "rendered": "Buscar",
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
